In [2]:
#importing basic libraries 
# Source - (Stack overflow, n.d.)
import numpy as np
import pandas as pd

In [3]:
#read excel file - metaClean43Brightspace
# Source - (Stack overflow, n.d.)
meta_clean = pd.read_excel(r"C:\Users\josel\OneDrive\Documents\Block 2\Buss stats\project\data set 2nd version\metaClean43Brightspace.xlsx")
df = pd.DataFrame(meta_clean)
df.head()

url                  title  \
0  https://www.metacritic.com/movie/!women-art-re...  !Women Art Revolution   
1  https://www.metacritic.com/movie/10-cloverfiel...    10 Cloverfield Lane   
2  https://www.metacritic.com/movie/10-items-or-less       10 Items or Less   
3          https://www.metacritic.com/movie/10-years               10 Years   
4  https://www.metacritic.com/movie/100-bloody-acres       100 Bloody Acres   

                     studio       rating  runtime  \
0       Hotwire Productions  | Not Rated     83.0   
1        Paramount Pictures      | PG-13    104.0   
2                Click Star          | R     82.0   
3  Anchor Bay Entertainment          | R    100.0   
4           Music Box Films  | Not Rated     91.0   

                                                cast              director  \
0                                                NaN  Lynn Hershman-Leeson   
1  John Gallagher Jr.,John Goodman,Mary Elizabeth...      Dan Trachtenberg   
2                 Jonah Hill,Morgan Freeman,Paz Vega       Brad Silberling   
3             Channing Tatum,Chris Pratt,Jenna Dewan          Jamie Linden   
4                                                NaN       Cameron Cairnes   

                                         genre  \
0                                  Documentary   
1  Action,Sci-Fi,Drama,Mystery,Thriller,Horror   
2                         Drama,Comedy,Romance   
3                         Drama,Comedy,Romance   
4                                Horror,Comedy   

                                             summary  \
0                                                NaN   
1  Waking up from a car accident, a young woman (...   
2  While researching a role as a supermarket mana...   
3                                                NaN   
4  Reg and Lindsay run an organic fertilizer busi...   

                                              awards  metascore  userscore  \
0                                                NaN         70        NaN   
1  #18MostDiscussedMovieof2016 , #1MostSharedMovi...         76        7.7   
2                                                NaN         54        5.8   
3                                                NaN         61        6.9   
4                                                NaN         63        7.5   

     RelDate  
0 2011-06-01  
1 2016-03-11  
2 2006-12-01  
3 2012-09-14  
4 2013-06-28

In [18]:
df.columns

Index(['url', 'title', 'studio', 'rating', 'runtime', 'cast', 'director',
       'genre', 'summary', 'awards', 'metascore', 'userscore', 'RelDate'],
      dtype='object')

In [4]:
# Display summary of the data frame, including amount of non-values, data types and memory usage, 
# Source - (Eddie, 2024)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11364 entries, 0 to 11363
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   url        11364 non-null  object        
 1   title      11364 non-null  object        
 2   studio     11014 non-null  object        
 3   rating     10297 non-null  object        
 4   runtime    11109 non-null  float64       
 5   cast       7662 non-null   object        
 6   director   11350 non-null  object        
 7   genre      11344 non-null  object        
 8   summary    5467 non-null   object        
 9   awards     4387 non-null   object        
 10  metascore  11364 non-null  int64         
 11  userscore  9259 non-null   float64       
 12  RelDate    11364 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(9)
memory usage: 1.1+ MB


Above it can be seen fromt he tabale above there are a total of 11364 row. As it can be seen column #2, 3, 4, 5, 6, 7, 8, 9 and 11 contains NaN values. 

In [5]:
#checking for null values in the data
# Source - (Eddie, 2024)
print(df.isnull().sum())

url             0
title           0
studio        350
rating       1067
runtime       255
cast         3702
director       14
genre          20
summary      5897
awards       6977
metascore       0
userscore    2105
RelDate         0
dtype: int64


In [6]:
# Calculate the percentage of NaN values for each column
# Source - (Nelamali, 2024)
NaN_Values_percentage = (df.isnull().sum() / len(df)) * 100

# The variable `NaN_Values_percentage` now contains the percentages
NaN_Values_percentage


url           0.000000
title         0.000000
studio        3.079901
rating        9.389300
runtime       2.243928
cast         32.576558
director      0.123196
genre         0.175994
summary      51.891939
awards       61.395635
metascore     0.000000
userscore    18.523407
RelDate       0.000000
dtype: float64

The percentages of NaN values in each column can be seen here, as these could still be usefull data. NaN values will be replaced with other values that are more suitable to each type of category. 

In [7]:
# NaN values for the studio column will be replaced or renamed as Unknown
# Source - (GeeksforGeeks, 2024)
df['studio'] = df['studio'].fillna('Unknown')

In [27]:
#verifying if NaN values were indeed changed for Unknown
# Source - Chatgpt
# Check if 'Unknown' exists in the 'studio' column and count occurrences
unknown_count = (df['studio'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'studio': {unknown_count}")

# Count the number of NaN values in the 'studio' column
nan_count_stud = df['studio'].isna().sum()
print(f"Number of NaN values in 'studio': {nan_count_stud}")


Number of 'Unknown' values in 'studio': 350
Number of NaN values in 'studio': 0


As it can be seen the NaN values are changed for Unknown, as the total NaN values were 350 before, now there are 0 NaN values and 350 values renamed as Unknown. 

In [10]:
# NaN values for the rating column will be replaced or renamed as '| Not Rated'
# Source - (GeeksforGeeks, 2024)
df['rating'] = df['rating'].fillna('| Not Rated')

In [28]:
#verifying if NaN values were indeed changed for '| Not Rated'
# Source - Chatgpt
# Check if '| Not Rated' exists in the 'rating' column and count occurrences
not_rated_count = (df['rating'] == '| Not Rated').sum()
print(f"Number of '| Not Rated' values in 'rating': {not_rated_count}")

# Count the number of NaN values in the 'rating' column
nan_count_rat = df['rating'].isna().sum()
print(f"Number of NaN values in 'rating': {nan_count_rat}")

Number of '| Not Rated' values in 'rating': 4027
Number of NaN values in 'rating': 0


The '| Not rated' values are more as there were existing row already named not rated. As it can be seen all the NaN values within teh rating column has been replaced for '| Not rated'.

In [ ]:
# NaN values for the cast column will be replaced or renamed as Unknown
# Source - (GeeksforGeeks, 2024)
df['cast'] = df['cast'].fillna('Unknown')

In [29]:
#verifying if NaN values were indeed changed for Unknown
# Source - Chatgpt
# Check if 'Unknown' exists in the 'cast' column and count occurrences
unknown_count = (df['cast'] == 'Unknown').sum()
print(f"Number of 'Unknown' values in 'cast': {unknown_count}")

# Count the number of NaN values in the 'cast' column
nan_count_cas = df['cast'].isna().sum()
print(f"Number of NaN values in 'cast': {nan_count_cas}")

Number of 'Unknown' values in 'cast': 3702
Number of NaN values in 'cast': 0


There are 0 NaN values now in the cast column.

In [ ]:
# NaN values for the summary column will be replaced or renamed as No Summary Available
# Source - (GeeksforGeeks, 2024)
df['summary'] = df['summary'].fillna('No Summary Available')

In [30]:
#verifying if NaN values were indeed changed for No Summary Available
# Source - Chatgpt
# Check if 'No Summary Available' exists in the 'summary' column and count occurrences
nosumav_count = (df['summary'] == 'No Summary Available').sum()
print(f"Number of 'No Summary Available' values in 'summary': {nosumav_count}")

# Count the number of NaN values in the 'cast' column
nan_count_sum = df['summary'].isna().sum()
print(f"Number of NaN values in 'summary': {nan_count_sum}")

Number of 'No Summary Available' values in 'summary': 5897
Number of NaN values in 'summary': 0


There are 0 NaN values now in the summary column.

In [16]:
# NaN values for the awards column will be replaced or renamed as No Awards
# Source - (GeeksforGeeks, 2024)
df['awards'] = df['awards'].fillna('No Awards')

In [31]:
#verifying if NaN values were indeed changed for No Awards
# Source - Chatgpt
# Check if 'No Awards' exists in the 'awards' column and count occurrences
no_awards_count = (df['awards'] == 'No Awards').sum()
print(f"Number of 'No Awards' values in 'awards': {no_awards_count}")

# Count the number of NaN values in the 'awards' column
nan_count_awa = df['awards'].isna().sum()
print(f"Number of NaN values in 'awards': {nan_count_awa}")

Number of 'No Awards' values in 'awards': 6977
Number of NaN values in 'awards': 0


As identified before the total NaN values were 6977, these NaN values are now replaced or renamed as No Awards

In [ ]:
# Find the lowest value in the 'userscore' column
# Source chatgpt
lowest_userscore = df['userscore'].min()
print(f"The lowest value in the 'userscore' column is: {lowest_userscore}")


The lowest value in the 'userscore' column is: 0.2


In [ ]:
# Check if there are any 0 values in the 'userscore' column and count them
# Source chatgpt
zero_count = (df['userscore'] == 0).sum()
if zero_count > 0:
    print(f"There are {zero_count} entries with a value of 0 in the 'userscore' column.")
else:
    print("There are no entries with a value of 0 in the 'userscore' column.")


There are no entries with a value of 0 in the 'userscore' column.


There are no userscores, with a value of 0 and teh lowest value is 0.2. Therfore, the NaN values will be replaced for a 0 value.

In [21]:
# NaN values for the userscore column will be replaced for 0
# Source - (GeeksforGeeks, 2024)
df['userscore'] = df['userscore'].fillna('0')

In [32]:
#verifying if NaN values were indeed changed for 0
# Source - Chatgpt
# Check if '0' exists in the 'userscore' column and count occurrences
zero_count = (df['userscore'] == '0').sum()
print(f"Number of '0' values in 'userscore': {zero_count}")

# Count the number of NaN values in the 'userscore' column
nan_count_scor = df['userscore'].isna().sum()
print(f"Number of NaN values in 'userscore': {nan_count_scor}")

Number of '0' values in 'userscore': 2105
Number of NaN values in 'userscore': 0


S previously identified the total number of NaN values are now replaced for 0. 

In [24]:
# Find the lowest value in the 'runtime' column
# Source chatgpt
lowest_runtime = df['runtime'].min()
print(f"The lowest value in the 'runtime' column is: {lowest_runtime}")

The lowest value in the 'runtime' column is: 21.0


In [25]:
# NaN values for the runtime column will be replaced for 0
# Source - (GeeksforGeeks, 2024)
df['runtime'] = df['runtime'].fillna('0')

In [33]:
#verifying if NaN values were indeed changed for 0
# Source - Chatgpt
# Check if '0' exists in the 'runtime' column and count occurrences
cero_count = (df['runtime'] == '0').sum()
print(f"Number of '0' values in 'runtime': {cero_count}")

# Count the number of NaN values in the 'runtime' column
nan_count_run = df['runtime'].isna().sum()
print(f"Number of NaN values in 'runtime': {nan_count_run}")

Number of '0' values in 'runtime': 255
Number of NaN values in 'runtime': 0


The total NaN values are now replaced with a 0

In [34]:
# Drop rows with NaN in 'director', and 'genre'
df = df[df['director'].notna()]
df = df[df['genre'].notna()]

In [35]:
# Count the number of NaN values in the 'genre' column
nan_count_gen = df['genre'].isna().sum()
print(f"Number of NaN values in 'genre': {nan_count_gen}")

Number of NaN values in 'genre': 0


In [36]:
# Count the number of NaN values in the 'director' column
nan_count_dir = df['director'].isna().sum()
print(f"Number of NaN values in 'director': {nan_count_dir}")

Number of NaN values in 'director': 0


The genre and director column does not contain any NaN values, the NaN values rows were succesfully removed.

In [38]:
# Rename the 'RelDate' column to 'reldate'
df.rename(columns={'RelDate': 'reldate'}, inplace=True)
df.head()

url                  title  \
0  https://www.metacritic.com/movie/!women-art-re...  !Women Art Revolution   
1  https://www.metacritic.com/movie/10-cloverfiel...    10 Cloverfield Lane   
2  https://www.metacritic.com/movie/10-items-or-less       10 Items or Less   
3          https://www.metacritic.com/movie/10-years               10 Years   
4  https://www.metacritic.com/movie/100-bloody-acres       100 Bloody Acres   

                     studio       rating runtime  \
0       Hotwire Productions  | Not Rated    83.0   
1        Paramount Pictures      | PG-13   104.0   
2                Click Star          | R    82.0   
3  Anchor Bay Entertainment          | R   100.0   
4           Music Box Films  | Not Rated    91.0   

                                                cast              director  \
0                                            Unknown  Lynn Hershman-Leeson   
1  John Gallagher Jr.,John Goodman,Mary Elizabeth...      Dan Trachtenberg   
2                 Jonah Hill,Morgan Freeman,Paz Vega       Brad Silberling   
3             Channing Tatum,Chris Pratt,Jenna Dewan          Jamie Linden   
4                                            Unknown       Cameron Cairnes   

                                         genre  \
0                                  Documentary   
1  Action,Sci-Fi,Drama,Mystery,Thriller,Horror   
2                         Drama,Comedy,Romance   
3                         Drama,Comedy,Romance   
4                                Horror,Comedy   

                                             summary  \
0                               No Summary Available   
1  Waking up from a car accident, a young woman (...   
2  While researching a role as a supermarket mana...   
3                               No Summary Available   
4  Reg and Lindsay run an organic fertilizer busi...   

                                              awards  metascore userscore  \
0                                          No Awards         70         0   
1  #18MostDiscussedMovieof2016 , #1MostSharedMovi...         76       7.7   
2                                          No Awards         54       5.8   
3                                          No Awards         61       6.9   
4                                          No Awards         63       7.5   

     reldate  
0 2011-06-01  
1 2016-03-11  
2 2006-12-01  
3 2012-09-14  
4 2013-06-28

The titles are now all in lower case letters.

# Reference

Eddie. (2024, October 14). How to handle missing data in Python? [Explained in 5 easy steps]. Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/05/dealing-with-missing-values-in-python-a-complete-guide/

Stack overflow. (n.d.). Clean up excel data with python using pandas package. Stack Overflow. Retrieved November 21, 2024, from https://stackoverflow.com/questions/42234523/clean-up-excel-data-with-python-using-pandas-package

Nelamali, N. (2024, October 22). Count NAN values in Pandas DataFrame. Spark by {Examples}. https://sparkbyexamples.com/pandas/count-nan-values-in-pandas/

GeeksforGeeks. (2024, August 21). Python | Pandas DataFrame.fillna() to replace Null values in dataframe. GeeksforGeeks. https://www.geeksforgeeks.org/python-pandas-dataframe-fillna-to-replace-null-values-in-dataframe/